### PR TITLE
ci/daint: add parallel steps to build with Intel/CRAY

### PR DESCRIPTION
**Pull request by dev-zero**

### CODE REVIEW

**Comment by dev-zero** [comment]
@shoshijak @alazzaro I've investigated a bit more on the Cray support in CMake and it seems that CMake has a specific feature for Cray and is perfectly capable to figure out the backend compiler. The only thing one has to do is to specify `-DCMAKE_SYSTEM_NAME=CrayLinuxEnvironment`.

After specifying that will `CMAKE_Fortran_COMPILER_ID` reflect the correct backend compiler instead of just Cray when it detects the Cray compiler wrappers (plus also reflect the systems settings when it comes to dynamic/static linking it seems).

This also means that we don't have to specify flags based on the compiler's executable name but can use CMake's `CMAKE_<LANG>_COMPILER_ID` to identify the compilers, which means we can also drop the explicit MPI compiler wrapper detection.

**Comment by dev-zero** [comment]
Ok, I've solved the build issues. As it seems does `nvcc` not play well with the Cray compiler (at least it is not aware of the flags for `CC` and passes by default flags `CC` does not accept), which is why a GCC compiler is needed when building with CUDA + Cray compiler.
Furthermore, GCC 4.8.5 seems to be broken for `libcusmm_benchmark.cu`: compilation takes more than 30 minutes (that's the time when I terminated the process forcefully) on Daint with the OS' GCC.

Now there are some warnings during compilation and some of the unittests fail:

```
[...]
ftn-3105 crayftn: WARNING MAKE_THREADS, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/dist/dbcsr_dist_methods.F, Line = 578, Column = 24 
  This call modifies loop control variable t.
[...]
ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 1704, Column = 21 
  This call modifies loop control variable irow3d.


ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 1704, Column = 21 
  This call modifies loop control variable icol3d.


ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 1706, Column = 21 
  This call modifies loop control variable icol3d.


ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 1706, Column = 21 
  This call modifies loop control variable irow3d.


ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 2154, Column = 24 
  This call modifies loop control variable icol3d.


ftn-3105 crayftn: WARNING MULTIPLY_3D, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/mm/dbcsr_mm_3d.F, Line = 2154, Column = 24 
  This call modifies loop control variable irow3d.
[...]
ftn-3105 crayftn: WARNING DBCSR_TAS_MERGE, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/tas/dbcsr_tas_reshape_ops.F, Line = 596, Column = 15 
  This call modifies loop control variable iproc.
[...]
ftn-3105 crayftn: WARNING DBCSR_COMPLETE_REDISTRIBUTE, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/ops/dbcsr_transformations.F, Line = 1800, Column = 24 
  This call modifies loop control variable dst_p.


ftn-3105 crayftn: WARNING DBCSR_REDISTRIBUTE, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/ops/dbcsr_transformations.F, Line = 2121, Column = 15 
  This call modifies loop control variable dst_p.


ftn-3105 crayftn: WARNING DBCSR_DATABLOCK_REDISTRIBUTE, File = ../../../scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray/src/ops/dbcsr_transformations.F, Line = 2392, Column = 18 
  This call modifies loop control variable dst_p.
[...]
```

the error log:

```
[...]
Test project /scratch/snx3000/jenkg90/jenkins-g90-DBCSR-115.cray

      Start  1: dbcsr_perf:inputs/test_H2O.perf

 1/17 Test  #1: dbcsr_perf:inputs/test_H2O.perf .......................***Failed    3.67 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f6180

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f6180

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f6180

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f6180

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.0



      Start  2: dbcsr_perf:inputs/test_rect1_dense.perf

 2/17 Test  #2: dbcsr_perf:inputs/test_rect1_dense.perf ...............***Failed    3.44 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.1



      Start  3: dbcsr_perf:inputs/test_rect1_sparse.perf

 3/17 Test  #3: dbcsr_perf:inputs/test_rect1_sparse.perf ..............***Failed    4.57 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.2

srun: error: nid04700: tasks 0-3: Aborted

srun: error: Task 0 reported exit for a second time.

srun: error: Task 1 reported exit for a second time.

srun: error: Task 2 reported exit for a second time.

srun: error: Task 3 reported exit for a second time.



      Start  4: dbcsr_perf:inputs/test_rect2_dense.perf

 4/17 Test  #4: dbcsr_perf:inputs/test_rect2_dense.perf ...............***Failed    3.72 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.3



      Start  5: dbcsr_perf:inputs/test_rect2_sparse.perf

 5/17 Test  #5: dbcsr_perf:inputs/test_rect2_sparse.perf ..............***Failed    3.55 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.4



      Start  6: dbcsr_perf:inputs/test_square_dense.perf

 6/17 Test  #6: dbcsr_perf:inputs/test_square_dense.perf ..............***Failed    3.28 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.5



      Start  7: dbcsr_perf:inputs/test_square_sparse.perf

 7/17 Test  #7: dbcsr_perf:inputs/test_square_sparse.perf .............***Failed    3.36 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.6



      Start  8: dbcsr_perf:inputs/test_square_sparse_bigblocks.perf

 8/17 Test  #8: dbcsr_perf:inputs/test_square_sparse_bigblocks.perf ...***Failed    3.60 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f61c0

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.7



      Start  9: dbcsr_unittest1

 9/17 Test  #9: dbcsr_unittest1 .......................................***Failed    3.69 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10dca80

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10dca80

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10dca80

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10dca80

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.8



      Start 10: dbcsr_unittest2

10/17 Test #10: dbcsr_unittest2 .......................................***Failed    3.55 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.9



      Start 11: dbcsr_unittest3

11/17 Test #11: dbcsr_unittest3 .......................................***Failed    3.78 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10c3980

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.10



      Start 12: dbcsr_tensor_unittest

12/17 Test #12: dbcsr_tensor_unittest .................................***Failed    3.58 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x1265980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x1265980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x1265980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x1265980

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.11



      Start 13: dbcsr_test_csr_conversions

13/17 Test #13: dbcsr_test_csr_conversions ............................***Failed    3.90 sec

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f0980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f0980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f0980

craylibs//google-perftools/src/tcmalloc.cc:644] Attempt to free invalid pointer: 0x10f0980

srun: error: nid04700: tasks 0-3: Aborted

srun: Terminating job step 14491865.12



      Start 14: libcusmm_unittest_multiply

14/17 Test #14: libcusmm_unittest_multiply ............................   Passed  767.32 sec

      Start 15: libcusmm_unittest_transpose

15/17 Test #15: libcusmm_unittest_transpose ...........................   Passed  281.59 sec

      Start 16: libcusmm_timer_multiply-autotuned

16/17 Test #16: libcusmm_timer_multiply-autotuned .....................   Passed  548.13 sec

      Start 17: libcusmm_timer_multiply-predicted

17/17 Test #17: libcusmm_timer_multiply-predicted .....................   Passed  607.67 sec
[...]
```

**Comment by shoshijak** [comment]
@dev-zero I came to the same conclusion as you about the build and was about to write you an email when I saw you had solved it already.

File: CMakeLists.txt [hunk 1]
```diff
@@ -88,24 +88,8 @@ endif ()
 
 # COMPILER CONFIGURATION:
 
-get_filename_component(Fortran_COMPILER_NAME ${CMAKE_Fortran_COMPILER} NAME)
```
Comment (shoshijak) [inline CMakeLists.txt]: Detection of the fortran compiler seems so much cleaner now, it's great that you could get rid of these lines.

**Comment by alazzaro** [comment]
Great job!
Yes, it better to use GCC with CUDA and your solution is perfect.

From my experience, CCE is really good to spot warnings with its static analyzer. Therefore I would suggest to add the flag:

`-h error_on_warning`

**Comment by alazzaro** [comment]
One more issue with the compiler: currently cmake does `-em -J.`, that produces modules in the current directory with name capital letters and extention .mod. I suggest to use

`-emf`

which produces modules in the same location of the .o files + all small letters.

**Comment by dev-zero** [comment]
interesting, I'll have to check why CMake's default is like that (not setting it like that explicitly)

**Comment by alazzaro** [comment]
It seems the compiler CCE 8.7.x is buggy, it calls a deallocate on this line:

https://github.com/cp2k/dbcsr/blob/517ae8af6bed4ecbcabb349c7eccfc631edfcb04/src/core/dbcsr_lib.F#L184

Actually, it crashes at the startup of the function. I cannot see anything wrong in the code, @dev-zero could you double-check?

**Comment by dev-zero** [comment]
there is some evidence that the issue we see is the one from here:

  https://bugzilla.redhat.com/show_bug.cgi?id=1569391

Checking the symbols of the linked `libtcmalloc_minimal.so.1` reveals that it doesn't export `aligned_alloc`, but `libnvrtc.so` seems to call it. So there is a chance that memory allocated via glibc's `aligned_alloc` is getting freed by tcmalloc's `free`, which triggers this kind of error message.

Actually, after the last CMake changes to add `-hsystem_alloc` and filtering `tcmalloc` (see comments in `CMakeLists.txt`) are the `Attempt to free invalid pointer` errors gone. One error remains:

```
[...]
      Start 12: dbcsr_tensor_unittest
12/17 Test #12: dbcsr_tensor_unittest .................................***Failed   25.10 sec
*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest': double free or corruption (!prev): 0x00000000025875c0 ***
*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest': corrupted double-linked list: 0x0000000009ee71d0 ***
======= Backtrace: =========
*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest': corrupted double-linked list: 0x0000000002535990 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x740ef)[0x2aaab24b50ef]
/lib64/libc.so.6(+0x740ef)[0x2aaab24b50ef]
/lib64/libc.so.6(+0x79646)[0x2aaab24ba646]
/lib64/libc.so.6(+0x79646)[0x2aaab24ba646]
/lib64/libc.so.6(+0x7a393)[0x2aaab24bb393]
/lib64/libc.so.6(+0x7a62e)[0x2aaab24bb62e]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(_DEALLOC+0x165)[0x2aaabb6ebe75]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(+0x18605)[0x2aaabb69c605]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(_DEALLOC+0x165)[0x2aaabb6ebe75]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(+0x18605)[0x2aaabb69c605]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(+0x18570)[0x2aaabb69c570]
/opt/cray/pe/cce/8.7.3/cce/x86_64/lib/libf.so.1(+0x18570)[0x2aaabb69c570]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x7d2ea1]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x7d0eee]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x6e05c0]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x69f451]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x40b165]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x7d2ea1]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x7d0eee]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x6e05c0]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x69f451]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x40b165]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x2aaab2461725]
/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-130.cray/tests/./dbcsr_tensor_unittest[0x40a769]
[...]
```

... I'm waiting for the test to finish, then will add another build with `-DCMAKE_BUILD_TYPE=Debug` to get debug symbols and hopefully a better stacktrace.

**Comment by hfp** [comment]
FYI, I am doing the same: https://github.com/hfp/libxsmm/blob/master/Makefile.inc#L698

**Comment by dev-zero** [comment]
@hfp I thought of libxsmm yesterday as well when I went through the linked bug report. I guess you also need aligned memory for some instructions?

**Comment by hfp** [comment]
@dev-zero No real need, but during my (still manual) validation of libxsmm on CRAY (at least for every release) I found some tests or samples crashing and came up with workaround (Git log says 27th of March 2017). At that time I was satisfied thinking an old/buggy version of TCMalloc may be the root cause, but your analysis sounds more thorough.

**Comment by dev-zero** [comment]
@alazzaro with `-g` on Cray we're much much slower. Should I revert it back to Debug or wait until we hear from CSCS/Cray?

**Comment by alazzaro** [comment]
@dev-zero Well, I think you can remove it now. The initial idea was to catch warnings to errors, but now we are doing the opposite... So, yes, we can use Release with the -M E...

**Comment by dev-zero** [comment]
Intel 19.0.1.144 works, Cray builds but still gives:

```
13/18 Test #13: dbcsr_tensor_unittest .................................***Failed    3.20 sec

*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-237.cray/tests/./dbcsr_tensor_unittest': double free or corruption (!prev): 0x000000000b5d3fc0 ***

*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-237.cray/tests/./dbcsr_tensor_unittest': double free or corruption (!prev): 0x000000000a90b100 ***

*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-237.cray/tests/./dbcsr_tensor_unittest': corrupted size vs. prev_size: 0x000000000a908350 ***

*** Error in `/scratch/snx3000/jenkg90/jenkins-g90-DBCSR-237.cray/tests/./dbcsr_tensor_unittest': corrupted size vs. prev_size: 0x000000000a90af10 ***
```

**Comment by dev-zero** [comment]
@alazzaro ready to be merged when tests succeed

**Comment by alazzaro** [comment]
So, in conclusion, Intel works, CCE 9.0.2 still crash at runtime... right?

**Comment by dev-zero** [comment]
Yes. And I've enabled Cray for compilation, but not tests. The other option would be to disable the tensor-test on Cray

**Pull request merged**